oxidation/crates/parsec_api_protocol/src/error.rs

```diff
@@ -0,0 +1,47 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+
+use parsec_api_types::DateTime;
+use thiserror::Error;
+
+use crate::ApiVersion;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum HandshakeError {
+    #[error("Handshake invalid answer (bad signature or challenge value)")]
+    FailedChallenge,
+    #[error("Handshake bad administration token")]
+    BadAdministrationToken,
+    #[error("Handshake bad identity")]
+    BadIdentity,
+    #[error("Handshake organization expired")]
+    OrganizationExpired,
+    #[error("Handshake rvk mismatch")]
+    RVKMismatch,
+    #[error("Handshake revoked device")]
+    RevokedDevice,
+    #[error("Handshake out of ballpark {0:x?}")]
+    OutOfBallpark(ChallengeDataReport),
+    #[error("No overlap between client API versions {client_versions:?} and backend API versions {backend_versions:?}")]
+    APIVersion {
+        client_versions: Vec<ApiVersion>,
+        backend_versions: Vec<ApiVersion>,
+    },
+    #[error("Handshake invalid message: {0}")]
+    InvalidMessage(String),
+}
+
+impl From<&'static str> for HandshakeError {
```

Comment:
I will prefere to use `HandshakeError::InvalidMessage(my_error.into())` directly in the caller instead of `my_error.into()`: the 1rst form is much more explicit while the 2nd encourage misleading behavior (typically when doing copy/paste when developing a new feature, one can be lead into thinking `my_error.into()` is the default way to create a `HandshakeError`, where it is only a special corner case)

Reply:
> it won't exist anymore when we remove them all

Then we should add a ```// TODO: temporary code to handle `&'static str` kind of errors```

Do you have an idea how complex it would be to clean the code from the need of this ?
I would naively guess the handshake is only used by the transport, which at this point is another unmerged PR, so better modify the transport PR than adding this legacy code handler.

> From trait allow the operator ? to auto convert from an error to an other error

That's precisely why this code is dangerous ;-)

Reply:
This is not dangerous, If you can't autoconvert that means you didn't handle correctly your error

Reply:
Naively you will want to remove this implementation, and it will find where you are doing autoconversions,
whereas if you don't do this auto conversion, you will have to change all the map_error

Reply:
The change was pretty trivial (see e975b6f03d607c1b46b5ec881141333e5e01a62b): replace `?` by `.map_err(|err| HandshakeError::InvalidMessage(err.into()))?` in a single file.

Adding temporary handler for this is not worth it ;-)